File: solutions/1143. Longest Common Subsequence/1143.py
# Time complexity: O(nm)
# Space complexity: O(nm)
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        n = len(text1)
        m = len(text2)
        dp = [[0] * (m + 1) for _ in range(n + 1)]

        for i in range(n):
            for j in range(m):
                dp[i + 1][j + 1] = \
                1 + dp[i][j] if text1[i] == text2[j] \
                else max(dp[i][j + 1], dp[i + 1][j])        

        return dp[n][m]  

File: solutions/1143. Longest Common Subsequence/1143.cpp
// Time complexity: O(nm)
// Space complexity: O(nm)
class Solution {
 public:
  int longestCommonSubsequence(string_view text1, string_view text2) {
    const int n = text1.length();
    const int m = text2.length();
    vector<vector<int>> dp(n + 1, vector<int>(m + 1));

    for (int i = 0; i < n; ++i)
      for (int j = 0; j < m; ++j)
        dp[i + 1][j + 1] = text1[i] == text2[j]
                               ? dp[i][j] + 1
                               : max(dp[i][j + 1], dp[i + 1][j]);

    return dp[n][m];
  }
};